cub/cub/detail/nvtx.cuh

```diff
@@ -0,0 +1,79 @@
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if __has_include(<source_location> )
+#  include <source_location>
+#endif // __has_include(<source_location>)
+
+// NVTX C++ wrapper documentation: https://nvidia.github.io/NVTX/doxygen-cpp/index.html
+#if __has_include(<nvtx3/nvtx3.hpp>) // TODO(bgruber): better check for the first CTK version shipping the header
+#  include <nvtx3/nvtx3.hpp>
+#else // __has_include(<nvtx3/nvtx3.hpp>)
+#  include "nvtx3.hpp"
+#endif // __has_include(<nvtx3/nvtx3.hpp>)
+
+CUB_NAMESPACE_BEGIN
+namespace detail
+{
+struct NVTXCCCLDomain
+{
+  static constexpr const char* name = "CCCL";
+};
+
+// FIXME(bgruber): check that no scope is introduced by NV_IF_TARGET!
+
+// Inserts a NVTX range starting here until the end of the current function scope
+#define CUB_NVTX_RANGE_SCOPE(name)                                                                                    \
+  NV_IF_TARGET(                                                                                                       \
+    NV_IS_HOST,                                                                                                       \
+    static const ::nvtx3::registered_string_in<CUB_NS_QUALIFIER::detail::NVTXCCCLDomain> __cub_nvtx3_func_name{name}; \
+    static const ::nvtx3::event_attributes __cub_nvtx3_func_attr{__cub_nvtx3_func_name};                              \
+    ::nvtx3::scoped_range_in<CUB_NS_QUALIFIER::detail::NVTXCCCLDomain> __cub_nvtx3_range{__cub_nvtx3_func_attr};      \
+    (void) __cub_nvtx3_range;)
```

Comment:
@miscco How can I conditionally enable this code only for host compilation, but not introduce a block scope, so the scope guard works?

Reply:
Discussed offline. We went for a `cuda::std::optional` that is only constructed in host code.